### DISCUSSION TITLE
Directus 9 Migration Patch

### DISCUSSION THREAD

**Discussion by erik-konrad** [directus/directus#3618] [Archive]
Hello dear community. We have made a patch for upgrading databases from Directus 8 to Directus 9, which we would like to share with the community.

You can find it here: https://github.com/erik-konrad/directus-9-migration-patch

Its a simple SQL file which you can patch in the database. There are some more steps to do manually. For this steps, look at the readme file.

If you found some issues (and more solutions), feel free to fork the project and start a pull request.

Best regards
Erik

**Comment by benhaynes** [comment]
Thanks Erik — we'll check this out internally and broadcast it to the community if everything looks good. We'll also be writing our own article for this migration process, but need to wrap up a few more internal things first. Really appreciate the contribution!!

**Comment by philleepflorence** [comment]
Oh wow...awesome!
Saves me the time of creating mine...
It took some time to transfer D6 to D8.
Hopefully the internal review will be positive.

**Comment by erik-konrad** [comment]
There is a new version of the migration patch. It now runs complete automatically with node.js. The only one thing to do is to change the config.json for your custom tables and db settings. The Directus Backend interface is running stable, all file references are now fully recovered and refactored from ID to UUID v4. The only one thing: you have to reconfigure your collection lists for backend roles and the permissions.

**Comment by erik-konrad** [comment]
@benhaynes dont use the old version for checking. the file references are not working correctly in it.

**Reply by arpijoy** [reply]
@erik-konrad I used your migration patch on Directus 8 schema. Unable to run directus9 rc60 due to the following:
CREATE INDEX directus_fields_group_foreign ON directus_fields (`group`); -- I see this in directus9 schema but ur migration-sql is commenting it out.
CREATE INDEX directus_fields_collection_foreign	 on directus_fields (collection);

23:14:44 ✨ Running migrations...
Error: ER_CANT_DROP_FIELD_OR_KEY: Can't DROP 'directus_fields_collection_foreign'; check that column/key exists
    at Query.Sequence._packetToError (/directus/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
    at Query.ErrorPacket (/directus/node_modules/mysql/lib/protocol/sequences/Query.js:79:18)
    at Protocol._parsePacket (/directus/node_modules/mysql/lib/protocol/Protocol.js:291:23)
    at Parser._parsePacket (/directus/node_modules/mysql/lib/protocol/Parser.js:433:10)
    at Parser.write (/directus/node_modules/mysql/lib/protocol/Parser.js:43:10)
    at Protocol.write (/directus/node_modules/mysql/lib/protocol/Protocol.js:38:16)
    at Socket.<anonymous> (/directus/node_modules/mysql/lib/Connection.js:88:28)
    at Socket.<anonymous> (/directus/node_modules/mysql/lib/Connection.js:526:10)
    at Socket.emit (events.js:315:20)
    at addChunk (internal/streams/readable.js:309:12)
    --------------------
    at Protocol._enqueue (/directus/node_modules/mysql/lib/protocol/Protocol.js:144:48)
    at Connection.query (/directus/node_modules/mysql/lib/Connection.js:198:25)
    at /directus/node_modules/knex/lib/dialects/mysql/index.js:126:18
    at new Promise (<anonymous>)
    at Client_MySQL._query (/directus/node_modules/knex/lib/dialects/mysql/index.js:120:12)
    at executeQuery (/directus/node_modules/knex/lib/execution/internal/query-executioner.js:37:17)
    at Client_MySQL.query (/directus/node_modules/knex/lib/client.js:134:12)
    at Runner.query (/directus/node_modules/knex/lib/execution/runner.js:130:36)
    at Runner.queryArray (/directus/node_modules/knex/lib/execution/runner.js:224:21)
    at ensureConnectionCallback (/directus/node_modules/knex/lib/execution/internal/ensure-connection-callback.js:11:19) {
  code: 'ER_CANT_DROP_FIELD_OR_KEY',
  errno: 1091,
  sqlMessage: "Can't DROP 'directus_fields_collection_foreign'; check that column/key exists",
  sqlState: '42000',
  index: 0,
  sql: 'alter table `directus_fields` drop foreign key `directus_fields_collection_foreign`'

**Reply by erik-konrad** [reply]
No, you are using directus-community/migration-tool. :-) Try it with mine, but dont forget to configure the junction tables and file references in the config.json. Then you have a working D9 DB. At this point, you can use the npx directus bootsrap function to run the db upgrade to the newest Directus version, but you have to skip the step with the indexes and constraints. For this, manually add the row in the directus_migration table and do the work manually.

**Reply by arpijoy** [reply]
@erik-konrad you are right, I was using the migration tool. However, that is giving me an exception as well. It is sending 2 payloads for 1 change. So I used my local environment to try out your tool. Couple of things if you can help with:
1. Which junction tables are you referring to?
2. I added manually the indexes and when I try to deploy D9 docker image on D8(migrated) DB, I get the above error

npx directus bootsrap function -> is this part of your toolkit or part of Directus9?

**Reply by erik-konrad** [reply]
I mean 
```
"oneFile": {
            "zbr_cases": "zbr_hero_image",
            "zbr_blogpost": "zbr_hero_image",
            "zbr_content_teaser": "zbr_teaser_image",
            "gastronomy": "logo|picture",
            "zbr_customer": "zbr_image",
            "zbr_employees": "zbr_image",
            "zbr_slider_items": "zbr_image"
 },        
"junctionTables": {
            "zbr_case_content_directus_files": "directus_files_id",
            "zbr_blog_content2directus_files": "directus_files_id"
}
```
this section in the config.json. This is for the file references. "oneFile" means the one file only file references in your custom tables. the key is the collection, the value is the field name. if you have multiple fields, split it with a pipe. like in gastronomy.
"junctionTables" is for multiple file references. at this point the key is the junction table and the value is the field in the junction table, which holds the id from your file.

this is needed, because in V8 the id from the files is int, and in v9 uuidv4.

after the script starts, there are some log information. there is a small bug, if the script is done, it does not stop. you have to do it with [ctrl] + c.

Then you have to reset your password and you get back the control about your v9 backend. Its possible that there are some smaller sql errors, like NOT NULL where it must be nullable, but these are easy to fix. Oh, and you have to reconfigure your interfaces in some cases.

the bootstrap function is a directus function to migrate small changes from minor patch to minor patch.

**Reply by arpijoy** [reply]
@erik-konrad I don't have any files in the system so I am concluding I do not need Junction tables

**Reply by erik-konrad** [reply]
then you can delete the content of the keys "oneFile" and "junctionTables". But the keys have to exist.

**Comment by arpijoy** [comment]
@erik-konrad this is excellent. Any updates from @benhaynes on if the migration patch looks good?

**Reply by benhaynes** [reply]
This is the official tool: ~~https://github.com/directus-community/extension-toolkit~~

Edit: Whooooops: https://github.com/directus-community/migration-tool

**Reply by arpijoy** [reply]
@benhaynes I was looking for the DB migration. What is this extension-toolkit for?

**Reply by benhaynes** [reply]
Migrating the DB. 🙄

**Reply by arpijoy** [reply]
@benhaynes the readme file suggests _custom Directus extensions in Typescript and Javascript_, what is the other repo https://github.com/directus-community/migration-tool for?

**Reply by benhaynes** [reply]
Sorry, pasted the wrong link! 😬

**Reply by arpijoy** [reply]
@benhaynes I am starting to work on the migration. Currently failing to fetch token: My post: https://directus-service.test.xxx.com/auth/authenticate sending a payload (POST) as recommended here: https://v8.docs.directus.io/api/authentication.html#tokens; getting error: {
    "error": {
        "code": 1,
        "message": "Not Found"
    }
}

**Comment by arpijoy** [comment]
@benhaynes After much issues and resolutions, ran the migration-tool which now has created errors when we try to create a collection. Sends 2 payloads at the same time. Not sure how to fix that
